locale/en/docs/guides/dont-block-the-event-loop.md

```diff
@@ -0,0 +1,135 @@
+
+So, you should definitely make sure you never block the Event Loop.
+In other words, each of your callbacks should complete quickly.
+This of course also applies to your await's, your Promise.then's, etc..
+
+A good way to think about this is to check the ["computational complexity"](https://en.wikipedia.org/wiki/Time_complexity) of your callbacks.
+If your callback takes a constant number of steps no matter what its arguments are, then you're definitely good to go.
+If your callback takes a different number of steps depending on its arguments, then you should think about how long the arguments might be.
+For example, code like `for (let i = 0; i < n; i++)` will run quickly for small `n` and much more slowly for large `n`.
+
+Luckily for us, Node uses Google's V8 engine for JavaScript.
+V8 is really fast, even for pretty large objects.
+If you might be doing double for-loops (`O(n^2)`) or sorting very large arrays (`O(nlogn)`), you should think about bounding the input, i.e. rejecting inputs that are too long.
+Otherwise, you're probably OK.
+
+### Regular expression Denial of Service (ReDoS)
+I said V8 was fast.
+I lied.
+There is one aspect of V8 that is actually pretty slow.
+[*Regular expressions*](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Regular_Expressions), while really handy, can actually take exponential-time.
+For example, if you are serving files and ensuring a path is valid, you might write "`if (f.match(/(\/.+)+$/) ...`" to make sure the path is a sequence of "/"-delimited names.
+Sadly, what if a client sends in the path `///.../\n` (say, 100 /'s followed by a newline character that the regexp's "." won't match)?
+Your server will take an extremely long time to evaluate the regexp with this input, and if the attacker adds another "/" the time taken will double.
+Try it!
+
+So why is this happening?
+The problem is that your server is using a "vulnerable regular expression", one that exposes you to Regular Expression Denial of Service ([ReDoS](https://www.owasp.org/index.php/Regular_expression_Denial_of_Service_-_ReDoS)).
+For this reason, you should be leery of using regular expressions to validate user input.
+
+There are some tools to check your regexps for safety, like
+- [safe-regex](https://github.com/substack/safe-regex)
+- [rxxr2](http://www.cs.bham.ac.uk/~hxt/research/rxxr2/)
+However, neither of these will catch all vulnerable regexps.
+
+A safer alternative is the [node-re2](https://github.com/uhop/node-re2) project, which uses Google's blazing-fast [RE2](https://github.com/google/re2) regexp engine. But be warned, RE2 is not 100% compatible with Node's regexps, so check for regressions if you swap in the node-re2 module to handle your regexps. And really complicated regexps may not be supported by node-re2.
+
+As usual, if you're trying to match something "obvious", like a URL or a file path, use an npm module!
```

Comment:
Quick question here, mostly for my own education: 
In that case, wouldn't `stringVar.indexOf(searchVar) !== -1` or `stringVar.includes(searchVar) !== false` be more efficient, and eliminating the need of another module?

I originally come from PHP, and replacing `preg_match()` with `stripos($string, $search) !== false` is mostly preferred for the same reasons, so I thought it would be the same in JS.

Reply:
If you're just doing a `strcmp` or `substr`-style string match, then yes, you should just use `indexOf`. You don't need a regular expression for this, nor will a regexp actually be dangerous in this context.

But sometimes people need more matching power for more complicated expressions, and a sufficiently complicated regexp can lead to REDOS.

Reply:
Okay, that's reassuring :smile: 
Maybe in that case you could specify this alternative, and that sufficiently complex regex are the real problem here? (I'm not quite sure what a sufficiently complete regex would be either; regex using escaped/reserved chars and/or logn regex?)

Reply:
Is the new REDOS section clearer? Still needs work?